python/etlv2/migration_job.py

```diff
@@ -0,0 +1,234 @@
+
+    return user_defined_types
+
+
+def register_all_udts(source_conn, dest_conn):
+    query = """
+    SELECT t.typname AS type_name, 
+           n.nspname AS schema_name,
+           array_agg(e.enumlabel ORDER BY e.enumsortorder) AS enum_labels
+    FROM pg_type t
+    JOIN pg_namespace n ON n.oid = t.typnamespace
+    JOIN pg_enum e ON t.oid = e.enumtypid
+    GROUP BY t.typname, n.nspname;
+    """
+
+    cursor = source_conn.cursor()
+    cursor.execute(query)
+    udt_info = cursor.fetchall()
+
+    for type_name, schema_name, enum_labels in udt_info:
+        full_type_name = f"{schema_name}.{type_name}"
+        
+        psycopg2.extensions.register_type(
+            psycopg2.extensions.new_type(
+                (source_conn.get_dsn_parameters()['dbname'] + '.' + full_type_name,),
+                full_type_name,
+                lambda value, cursor: value
+            )
+        )
+    
+    cursor.close()
+
+    # Register the UDTs in the destination database
+    for type_name, schema_name, enum_labels in udt_info:
+        full_type_name = f"{type_name}"
+        enum_lst = ', '.join([f"'{label}'" for label in enum_labels])
+        
+        create_enum_type_query = f"CREATE TYPE {full_type_name} AS ENUM ({enum_lst});"
+        udt_query = f"""
+                    DO $$
+                    BEGIN
+                        IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = '{type_name}') THEN
+                        {create_enum_type_query}
+                        END IF;
+                    END$$;
+                    """
+        dest_cursor = dest_conn.cursor()
+        dest_cursor.execute(udt_query)
+    dest_cursor.close()
+
+# only fetch columns that are present in the source table
+# transform the data fetched 
+def fetch_and_transform(conn, table_name: str, src_table_columns: List[str], dest_table_columns: List[str], batch_size = 10 ** 6) -> List[tuple[Any, ...]]:
+    """
+        Only fetch columns that are present in BOTH the source table and the dest table
+        For columns that only exists in the dest table but not in the source table, 
+        we transform the data fetched from the source table to fit the dest table's schema by appending NULL values.
+    """
+    assert table_name in TABLES, "you must specify the tables you want to fetch in TABLES"
+    res_data = []
+    request_columns = [col for col in dest_table_columns if col in src_table_columns] # all columns that exist both in the src table and dest table
+    with conn.cursor() as cursor:
+        request_columns_str = ", ".join(request_columns) if request_columns else "*"
+        query = f"SELECT {request_columns_str} FROM {table_name};"
+        if entity_ids: # use the specified entity id as filter condition
+            query = query[:-1] + " WHERE "
+            filter_column = "id" if table_name == "entity" else "organization_id" if table_name == "planter" else "planter_id"
+            entity_ids_str = ", ".join(map(str, entity_ids))
```

Comment:
@1nzexi I can not understand logic here, how we get ids here for filter? I didn't find the place to update this entity_ids, where the data from?